Sources/MongoSwift/BSON/BSON.swift

```diff
@@ -0,0 +1,413 @@
+import Foundation
+
+/// Enum representing a BSON value.
+/// - SeeAlso: bsonspec.org
+public enum BSON {
+    /// A BSON double.
+    case double(Double)
+
+    /// A BSON string.
+    /// - SeeAlso: https://docs.mongodb.com/manual/reference/bson-types/#string
+    case string(String)
+
+    /// A BSON document.
+    case document(Document)
+
+    /// A BSON array.
+    indirect case array([BSON])
+
+    /// A BSON binary.
+    case binary(Binary)
+
+    /// A BSON undefined.
+    case undefined
+
+    /// A BSON ObjectId.
+    /// - SeeAlso: https://docs.mongodb.com/manual/reference/bson-types/#objectid
+    case objectId(ObjectId)
+
+    /// A BSON boolean.
+    case bool(Bool)
+
+    /// A BSON datetime.
+    /// - SeeAlso: https://docs.mongodb.com/manual/reference/bson-types/#date
+    case date(Date)
+
+    /// A BSON null.
+    case null
+
+    /// A BSON regular expression.
+    case regex(RegularExpression)
+
+    /// A BSON dbPointer.
+    case dbPointer(DBPointer)
+
+    /// A BSON symbol.
+    case symbol(Symbol)
+
+    /// A BSON code.
+    case code(Code)
+
+    /// A BSON code with scope.
+    case codeWithScope(CodeWithScope)
+
+    /// A BSON int32.
+    case int32(Int32)
+
+    /// A BSON timestamp.
+    /// - SeeAlso: https://docs.mongodb.com/manual/reference/bson-types/#timestamps
+    case timestamp(Timestamp)
+
+    /// A BSON int64.
+    case int64(Int64)
+
+    /// A BSON decimal128.
+    case decimal128(Decimal128)
+
+    /// A BSON minKey.
+    case minKey
+
+    /// A BSON maxKey.
+    case maxKey
+
+    /// If this `BSON` is an `.int32`, return it as an `Int32`. Otherwise, return nil.
+    public var int32Value: Int32? {
```

Comment:
these are the helpers that simply extract a value directly if possible with no coercion. There are helpers like this for types that have associated values (e.g. `.int32(Int32)`). For types like `.null` or `.maxKey`, a user can just do `if bson == .null { ... }`, so theres no real point in adding helpers.

Reply:
I think it's `if case .null = bson`, but in any case (lol) I agree helper isn't really valuable

Reply:
`==` also works here due to the `Equatable` conformance, but yeah that's another option thats pretty much equally simple. The benefit of using `Equatable` though is that you get the negation `!=`, and you can use it outside of control flow situations, e.g. `let x = BSON.null == BSON.string("not equal")`

Reply:
oooo right, good point